### PR TITLE
Feat(web-react): Collapse react implementation #DS-275

### PR DESCRIPTION
**Pull request by tomassychra**
# Jira issue [DS-275](https://jira.lmc.cz/browse/DS-275)

Implementation of Modal component for REACT

## 👉 What was done

1. Component & documentation

### CODE REVIEW

**Review by literat — APPROVED** [review]

File: packages/web-react/src/components/Collapse/useCollapse.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { useState } from 'react';
```
Comment (literat) [inline packages/web-react/src/components/Collapse/useCollapse.ts:1]: I am missing test for this file.

File: packages/web-react/src/components/Collapse/useCollapseStyleProps.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import classNames from 'classnames';
```
Comment (literat) [inline packages/web-react/src/components/Collapse/useCollapseStyleProps.ts:1]: I am missing test for this file.

File: packages/web-react/src/components/Collapse/useResizeHeight.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { useState, useEffect, MutableRefObject } from 'react';
```
Comment (literat) [inline packages/web-react/src/components/Collapse/useResizeHeight.ts:1]: Please add a test for this if it is possible.

File: packages/web-react/src/components/Collapse/useCollapseAriaProps.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { useRef, Ref, CSSProperties, MutableRefObject } from 'react';
```
Comment (literat) [inline packages/web-react/src/components/Collapse/useCollapseAriaProps.ts:1]: Please add a test for this if it is possible.

File: packages/web-react/src/types/collapse.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import { ChildrenProps, ClickEvent, StyleProps, Booleanish } from './shared';
+
+export type CollapseElementType = 'div' | 'article' | 'section' | 'main' | 'header' | 'footer' | 'span';
+
+export type CollapseResponsiveType = undefined | 'mobile' | 'tablet' | 'desktop';
+
+export type CollapseTriggerProps = {
+  onClick: (event: ClickEvent) => void;
+  className?: string | undefined;
+  'aria-expanded': Booleanish | undefined;
```
Comment (literat) [inline packages/web-react/src/types/collapse.ts:11]: Is this ok? I think that `undefined` => `false`. So we should handle somewhere the right value.

File: packages/web-react/src/components/Collapse/README.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+## CollapseRenderProps
+
+| Prop name                  | Type      | Description                |
+| -------------------------- | --------- | -------------------------- |
+| `collapsed`                | `boolean` | When collapse is collapsed |
+| `trigger`                  | `Object`  | Trigger properties         |
+| `trigger.onClick`          | `string`  | Trigger onClick event      |
+| `trigger.className`        | `string`  | Trigger state classname    |
+| `trigger['aria-expanded']` | `string`  | Trigger aria expanded      |
```
Comment (literat) [inline packages/web-react/src/components/Collapse/README.md:59]: I think that `aria-expanded` should be handled based on `isCollapsed` and should not be accessible. What do you think?

File: packages/web-react/src/types/shared/index.ts [hunk 7]
```diff
@@ -17,3 +17,5 @@ export enum SpiritBreakpoints {
   TABLET = 'tablet',
   DESKTOP = 'desktop',
 }
+
+export type Booleanish = boolean | 'true' | 'false';
```
Comment (literat) [inline packages/web-react/src/types/shared/index.ts:21]: This seems not right to me 🤔 .

**Review by literat — CHANGES_REQUESTED** [review]

File: packages/web-react/src/components/Collapse/useCollapseStyleProps.ts (hunk 2 above)
Comment (tomassychra) [reply packages/web-react/src/components/Collapse/useCollapseStyleProps.ts:1]: Done

File: packages/web-react/src/types/shared/index.ts (hunk 7 above)
Comment (tomassychra) [reply packages/web-react/src/types/shared/index.ts:21]: This is actual type for all `aria-*` props which have value 'true' or 'false' is called `Booleanish` and this is the type. Unfortunately is not exported and you must create it manually.

File: packages/web-react/src/components/Collapse/useCollapse.ts (hunk 1 above)
Comment (tomassychra) [reply packages/web-react/src/components/Collapse/useCollapse.ts:1]: Done

File: packages/web-react/src/components/Collapse/useResizeHeight.ts (hunk 3 above)
Comment (tomassychra) [reply packages/web-react/src/components/Collapse/useResizeHeight.ts:1]: Done

File: packages/web-react/src/components/Collapse/useCollapseAriaProps.ts (hunk 4 above)
Comment (tomassychra) [reply packages/web-react/src/components/Collapse/useCollapseAriaProps.ts:1]: Done

File: packages/web-react/src/types/collapse.ts (hunk 5 above)
Comment (tomassychra) [reply packages/web-react/src/types/collapse.ts:11]: Good point, i set it to `Booleanish` only, so it includes boolean already

File: packages/web-react/src/components/Collapse/README.md (hunk 6 above)
Comment (tomassychra) [reply packages/web-react/src/components/Collapse/README.md:59]: These props are **CollapseRenderProps**, which you get when rendering the button, so I think it's good to return it, as well as the `aria-controls`.

File: packages/web-react/src/components/Collapse/useCollapseAriaProps.ts [hunk 8]
```diff
@@ -49,7 +49,7 @@ export const useCollapseAriaProps = (props: CollapseAriaPropsProps): CollapseAri
   const triggerProps = {
     [NAME_ARIA_EXPANDED]: String(isCollapsed) as Booleanish,
     [NAME_ARIA_CONTROLS]: String(id),
-    onClick: toggleHandler,
+    onClick: toggleHandler as (event: ClickEvent) => void,
```
Comment (literat) [inline packages/web-react/src/components/Collapse/useCollapseAriaProps.ts:52]: Is there some problem with the type that you must enforce the type for this handler?

Comment (literat) [reply packages/web-react/src/components/Collapse/useCollapseAriaProps.ts:52]: Ha, I test it and see the problem with mismatched type of aria-expanded, which is one time `string` and other time `Booleanish`.

```git
diff --git packages/web-react/src/components/Collapse/__tests__/useCollapseAriaProps.test.ts packages/web-react/src/components/Collapse/__tests__/useCollapseAriaProps.test.ts
index bc1fe40f..fbc153db 100644
--- packages/web-react/src/components/Collapse/__tests__/useCollapseAriaProps.test.ts
+++ packages/web-react/src/components/Collapse/__tests__/useCollapseAriaProps.test.ts
@@ -6,6 +6,7 @@ describe('useCollapseAriaProps', () => {
     const props = {
       id: 'test-collapse-id',
       isCollapsed: true,
+      toggleHandler: () => null,
     };
     const { result } = renderHook(() => useCollapseAriaProps(props));
 
diff --git packages/web-react/src/components/Collapse/useCollapseAriaProps.ts packages/web-react/src/components/Collapse/useCollapseAriaProps.ts
index 8f6c10ac..bb9248ed 100644
--- packages/web-react/src/components/Collapse/useCollapseAriaProps.ts
+++ packages/web-react/src/components/Collapse/useCollapseAriaProps.ts
@@ -10,7 +10,7 @@ export interface CollapseAriaPropsProps {
   id: string;
   isCollapsed: boolean;
   collapsibleToBreakpoint?: CollapseResponsiveType;
-  toggleHandler?: (event: ClickEvent) => void;
+  toggleHandler: (event: ClickEvent) => void;
 }
 
 export interface CollapseAriaPropsReturn {
@@ -47,9 +47,9 @@ export const useCollapseAriaProps = (props: CollapseAriaPropsProps): CollapseAri
     ref: contentElementRef,
   };
   const triggerProps = {
-    [NAME_ARIA_EXPANDED]: String(isCollapsed) as Booleanish,
+    [NAME_ARIA_EXPANDED]: isCollapsed,
     [NAME_ARIA_CONTROLS]: String(id),
-    onClick: toggleHandler as (event: ClickEvent) => void,
+    onClick: toggleHandler,
   };
 
   return {
(END)
```

**Review by literat — APPROVED** [review]
Well done. 👏  Few minor issues and you are ready-to-merge.

**Pull request merged**